### PR TITLE
THREESCALE-10238 - mapping rules reconciliation improvement

### PR DESCRIPTION
**Pull request by valerymo**
## CR will be closed. Please see PR https://github.com/3scale/3scale-operator/pull/924  - suggested for review and approval.
Jira: https://issues.redhat.com/browse/THREESCALE-10238

- Please see detailed description and related tasks in [THREESCALE-10238](https://issues.redhat.com/browse/THREESCALE-10238)

# Example of the problem that fixed in current PR

- **Product CR contains "duplicated" rules:**
productCRDUplicatedRules

## Mapping Rules in portal Before Fix 
beforeFix

## Mapping Rules in portal After Fix in current PR
afterFix

## Add new Rule in CR:
Screenshot from 2024-01-25 10-09-11

Screenshot from 2024-01-25 10-08-33

##  Problem to be resolved  -TODO ?
- **can't add "duplicated" rule in CR** . 
 Can't add rule if identical rule exists. Although rule can be edited (increment field can be changed)
Need investigate the problem and allow this.

- when upgrade to new 3scale Release (with PR)  - Not valid Old rules remains and need to be cleaned manually (?)
problemAferFixExistedRules

 ## Validation 
 <details>

You can try compare the behavior on muster branch (before fix), and after fix (current PR)

```
make install
oc new-project 3scale-test
make download
make run
```

- On another window:
Apply following files. The examples for backend and product - provided below. Other - please ask me if required.
```
oc apply -f s3-creds-secret.yaml
oc apply -f apimanagerCR.yaml
oc apply -f backend.yml
oc apply -f product.yml
```
- Check mapping rules, compare with Product and Backend CRs
- Try add rules (uniq and identical)

### Product and Backend files examples
file contains duplicated rules to show the problem

```yaml
apiVersion: capabilities.3scale.net/v1beta1
kind: Product
metadata:
  name: product1
spec:
  name: "OperatedProduct 1"
  systemName: "operatedproduct1"
  description: "Operated Product 1 for reconciliation testing"
  backendUsages:
    backend1:
      path: /
  mappingRules:
    - httpMethod: GET
      increment: 1
      metricMethodRef: hits
      pattern: /pets      
    - httpMethod: GET
      increment: 1
      metricMethodRef: hits
      pattern: /pets      
    - httpMethod: GET
      increment: 1
      metricMethodRef: hits
      pattern: /pets      
    - httpMethod: GET
      increment: 1
      metricMethodRef: method01
      pattern: /pets 
      
  metrics:
    hits:
      description: Number of API hits
      friendlyName: Hits
      unit: hit
  methods:
    method01:
      friendlyName: Method01
```

```yaml
apiVersion: capabilities.3scale.net/v1beta1
kind: Backend
metadata:
  name: backend1
spec:
  name: "Operated Backend 1"
  systemName: "backend1"
  description: "Operated Backend 1 for reconciliation testing"
  privateBaseURL: "https://example.3scale.net"
  mappingRules:
    - httpMethod: PUT
      increment: 1
      metricMethodRef: hits
      pattern: /pets
    - httpMethod: GET
      increment: 1
      metricMethodRef: hits
      pattern: /pets      
    - httpMethod: GET
      increment: 1
      metricMethodRef: hits
      pattern: /cars
    - httpMethod: GET
      increment: 1
      metricMethodRef: method01
      pattern: /pets
    - httpMethod: GET
      increment: 1
      metricMethodRef: method01
      pattern: /pets
  metrics:
    hits:
      description: Number of API hits
      friendlyName: Hits
      unit: hit
  methods:
    method01:
      friendlyName: Method01

```

 </details>

### CODE REVIEW

File: pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl
Comment (briangallagher) [inline pkg/assets/assets/monitoring/system-grafana-dashboard-1.json.tpl:1]: @valerymo Why is this file included?

File: pkg/assets/bindata.go
Comment (briangallagher) [inline pkg/assets/bindata.go:1]: @valerymo Why is this file included?

File: doc/openapi-user-guide.md
Comment (briangallagher) [inline doc/openapi-user-guide.md:1]: @valerymo Why is this file included?

File: controllers/capabilities/mapping_rules.go [hunk 1]
```diff
@@ -105,9 +114,12 @@ func (t *ProductThreescaleReconciler) syncMappingRules(_ interface{}) error {
 
 func (t *ProductThreescaleReconciler) processNotDesiredMappingRules(notDesiredList []threescaleapi.MappingRuleItem) error {
 	for _, mappingRule := range notDesiredList {
-		err := t.productEntity.DeleteMappingRule(mappingRule.ID)
-		if err != nil {
-			return err
+		if mappingRule.HTTPMethod == "" || mappingRule.Pattern == "" || mappingRule.Pattern == "/" {
```
Comment (MStokluska) [inline controllers/capabilities/mapping_rules.go:117]: notDesiredMapping rule is not just a rule without http method, pattern or pattern = /
It's also rules that exist in 3scale but do not exist on the CR. Please revert this change.

File: controllers/capabilities/mapping_rules.go [hunk 2]
```diff
@@ -105,9 +114,12 @@ func (t *ProductThreescaleReconciler) syncMappingRules(_ interface{}) error {
...
+			t.logger.V(1).Info("processNotDesiredMappingRules, delete", "HTTPMethod", mappingRule.HTTPMethod, "Pattern", mappingRule.Pattern, "MetricID", strconv.FormatInt(mappingRule.MetricID, 10))
```
Comment (MStokluska) [inline controllers/capabilities/mapping_rules.go:118]: As suggested on our chat, I would not go into metricID here at all. Instead, pull the mappingRule.Position.

File: controllers/capabilities/mapping_rules.go [hunk 3]
```diff
@@ -120,28 +132,19 @@ func (t *ProductThreescaleReconciler) getExistingMappingRules() (map[string]thre
 		return nil, fmt.Errorf("Error getting product [%s] mappingrules: %w", t.resource.Spec.SystemName, err)
 	}
 	for _, item := range existingList.MappingRules {
-		key := fmt.Sprintf("%s:%s", item.Element.HTTPMethod, item.Element.Pattern)
+		key := fmt.Sprintf("%s:%s:%s", item.Element.HTTPMethod, item.Element.Pattern, strconv.FormatInt(item.Element.MetricID, 10))
```
Comment (MStokluska) [inline controllers/capabilities/mapping_rules.go:135]: Going by position in general is better.
When ranging through mapping rules from CR, we always go top down (top being 0, so position+1)
When ranging through mapping rules from 3scale, we can use item.Element.Position to get position.

File: controllers/capabilities/proxy.go [hunk 4]
```diff
@@ -197,20 +196,13 @@ func (t *ProductThreescaleReconciler) syncProxyOIDC(params threescaleapi.Params,
...
 	issuerEndpoint := oidcSpec.IssuerEndpoint
 	if issuerEndpoint == "" {
 		if oidcSpec.IssuerEndpointRef == nil {
 			// If missing both IssuerEndpoint and  IssuerEndpointRef in OpenApi CR - Product will fail SyncProxy
-			fieldErrors = append(fieldErrors, field.Invalid(oidcRefFldPath.Child("IssuerEndpoint"), t.resource.Spec.OIDCSpec().IssuerEndpoint, "no IssuerEndpoint nor IssuerEndpointRef found in OIDC spec in CR. Product OpenID Connect Issuer will not be set"))
-			return &helper.SpecFieldError{
-				ErrorType:      helper.InvalidError,
-				FieldErrorList: fieldErrors,
-			}
+			return fmt.Errorf("missing IssuerEndpoint definition in OIDC spec in openapi CR. Product OpenID Connect Issuer will not be set.")
```
Comment (MStokluska) [inline controllers/capabilities/proxy.go:205]: please rebase Valery.

**Comment by valerymo** [comment]
New PR https://github.com/3scale/3scale-operator/pull/924  - with updated solution is suggested for review and approval.
Close current PR without merge.   
New PR is addressing Mapping rules issues. The only "limitation" behavior - Rules added via UI will be removed in next reconciliation cicle. If this limitation will be found important - suggest to revise  currect solution + add counter for rules to distingush between identical rules in UI and CR.
